Improve name inference API by introducing a TensorName helper struct

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28904 Improve name inference API by introducing a TensorName helper struct**

Motivation
============

Before this PR, a core problem with writing name inference rules was
that each rule needed to handle misalignment by themselves. A misaligned
name occurs when we are matching None with a non-None name, but the
non-None name already exists in the first tensor.

For example, `A` is misaligned in `Tensor[A, None] + Tensor[None, A]`.

Each op handled this in a custom way
- align_from_right (used by broadcasting) handles misalignment
- compute_matmul_outnames checks for misalignment across batch and
feature dimensions.

We can actually codify "misalignment" into something more rigorous by
folding it into the definition of `match` and eliminate special handling
of "misalignment". That is what this PR attempts to do.

Approach
============

Definition: Two names in two tensors *match* if they are equal, or if at
least one of them is a wildcard that can be *refined* to the other name.

With this new definition, to check if two names match, we need to know
about the names list that each name came from to determine if a wildcard
can successfully be *refined* to the other name.

For example, consider the following:
```
tensor: Tensor[A, None]
other: Tensor[None, A]`
```
when unifying `tensor.names[-1]` with `other.names[-1]`, we see that
`tensor.names[-1]` is None and `other.names[-1]` is A. Then we check to
see if `tensor.names[-1]` can be refined to `A`; it can't be refined if
there is already an `A` in `tensor.names`.

Enter `TensorNames`.
A TensorName represents a Dimname associated with some DimnameList
(that came from a Tensor).

`TensorNames` is a list of such TensorName objects with some helper
functions attached.

One can perform the following operations:
- unify two `TensorName` objects
- unify two `TensorNames` objects with right alignment.

Plan
============

This PR changes `compute_matmul_outnames` to use `TensorNames` to
demonstrate how they make writing name inference rules easier. In the
future I'll convert other name inference rules to use `TensorNames` as
well.

Test Plan
- run all tests

Differential Revision: [D18270666](https://our.internmc.facebook.com/intern/diff/D18270666)

zou3519 committed:
Improve name inference API by introducing a TensorName helper struct

Motivation
^^^^^^^^^^

Before this PR, a core problem with writing name inference rules was
that each rule needed to handle misalignment by themselves. A misaligned
name occurs when we are matching None with a non-None name, but the
non-None name already exists in the first tensor.

For example, `A` is misaligned in `Tensor[A, None] + Tensor[None, A]`.

Each op handled this in a custom way
- align_from_right (used by broadcasting) handles misalignment
- compute_matmul_outnames checks for misalignment across batch and
feature dimensions.

We can actually codify "misalignment" into something more rigorous by
folding it into the definition of `match` and eliminate special handling
of "misalignment". That is what this PR attempts to do.

Approach
^^^^^^^^

Definition: Two names in two tensors *match* if they are equal, or if at
least one of them is a wildcard that can be *refined* to the other name.

With this new definition, to check if two names match, we need to know
about the names list that each name came from to determine if a wildcard
can successfully be *refined* to the other name.

Enter `TensorNames`.
A TensorName represents a Dimname associated with some DimnameList
(that came from a Tensor).

`TensorNames` is a list of such TensorName objects with some helper
functions attached.

One can perform the following operations:
- unify two `TensorName` objects
- unify two `TensorNames` objects with right alignment.

Plan
^^^^

This PR changes `compute_matmul_outnames` to use `TensorNames` to
demonstrate how they make writing name inference rules easier. In the
future I'll convert other name inference rules to use `TensorNames` as
well.

Test Plan
- run all tests

[ghstack-poisoned]

zou3519 committed:
Update on "Improve name inference API by introducing a TensorName helper struct"


Motivation
============

Before this PR, a core problem with writing name inference rules was
that each rule needed to handle misalignment by themselves. A misaligned
name occurs when we are matching None with a non-None name, but the
non-None name already exists in the first tensor.

For example, `A` is misaligned in `Tensor[A, None] + Tensor[None, A]`.

Each op handled this in a custom way
- align_from_right (used by broadcasting) handles misalignment
- compute_matmul_outnames checks for misalignment across batch and
feature dimensions.

We can actually codify "misalignment" into something more rigorous by
folding it into the definition of `match` and eliminate special handling
of "misalignment". That is what this PR attempts to do.

Approach
============

Definition: Two names in two tensors *match* if they are equal, or if at
least one of them is a wildcard that can be *refined* to the other name.

With this new definition, to check if two names match, we need to know
about the names list that each name came from to determine if a wildcard
can successfully be *refined* to the other name.

For example, consider the following:
```
tensor: Tensor[A, None]
other: Tensor[None, A]`
```
when unifying `tensor.names[-1]` with `other.names[-1]`, we see that
`tensor.names[-1]` is None and `other.names[-1]` is A. Then we check to
see if `tensor.names[-1]` can be refined to `A`; it can't be refined if
there is already an `A` in `tensor.names`.

Enter `TensorNames`.
A TensorName represents a Dimname associated with some DimnameList
(that came from a Tensor).

`TensorNames` is a list of such TensorName objects with some helper
functions attached.

One can perform the following operations:
- unify two `TensorName` objects
- unify two `TensorNames` objects with right alignment.

Plan
============

This PR changes `compute_matmul_outnames` to use `TensorNames` to
demonstrate how they make writing name inference rules easier. In the
future I'll convert other name inference rules to use `TensorNames` as
well.

Test Plan
- run all tests

[ghstack-poisoned]

nairbv commented on aten/src/ATen/TensorNames.h:
```diff
@@ -0,0 +1,64 @@
+  explicit TensorName(ArrayRef<Dimname> origin, int origin_idx)
+    : origin_(origin),
+      origin_idx_(origin_idx),
+      name_(origin[maybe_wrap_dim(origin_idx, origin.size())]) {}
```
it's not directly related to this PR but at some point we should add better comments for maybe_wrap_dim.

zou3519 replied:
What sort of comments did you want to see for maybe_wrap_dim? Just describing what it does?

nairbv replied:
right, I think I've re-read that code a few times come across the function. When I see "wrap" I think composition and get confused so go check the definition, but there's no comment and the code in WrapDimMinimal.h takes a bit of thinking to understand.

zou3519 replied:
> When I see "wrap" I think composition

Out of curiosity, what do you mean by this? Something like monads?

nairbv replied:
I just mean when you 'wrap' some object in another object (making it a member) as in:
https://en.wikipedia.org/wiki/Adapter_pattern
https://en.wikipedia.org/wiki/Composition_over_inheritance
Or when we wrap some function call by calling it from some other function:
https://en.wikipedia.org/wiki/Wrapper_function

In some sense this particular PR can be thought of as wrapping dimname/dimnamelist instances in a class that adds extra functionality, which I think contributes to me not immediately jumping to the other meaning of index wrapping when I see this function call.

nairbv replied:
Actually you refer to exactly what I meant, above:
// TensorName and TensorNames are wrappers around Dimname and DimnameList
that meaning being different from the `maybe_wrap_dim` meaning of 'wrap'.

nairbv commented on aten/src/ATen/TensorNames.cpp:
```diff
@@ -0,0 +1,103 @@
+  }
+}
+
+TensorNames TensorNames::unifyFromRight(const TensorNames& other, const char* op_name) const {
```
eventually this should effectively replace at::unify_from_right, right? should we have equivalent unit tests for these functions in NamedTensor_test.cpp?

zou3519 replied:
It should yes. Yes, equivalent unit tests would be great, thanks for pointing that out. I'll add them into this PR

zou3519 replied:
Done

nairbv commented on aten/src/ATen/TensorNames.h:
```diff
@@ -0,0 +1,64 @@
+// None (in tensor) cannot match A (in other) because if the None were refined
+// to A, `tensor` would have duplicate names [A, A]. Therefore we need to check
+// tensor.names [A, None] for the existence of A.
+struct CAFFE2_API TensorName {
```
It might start getting confusing how we have NamedTensorMeta, TensorName/DimName and TensorNames/DimNameList. Maybe this could be a TensorNamesUnifier or... not sure if there's some better name.

zou3519 replied:
I'm definitely taking better name suggestions.

TensorName is actually "namedinference::TensorName", and Dimname is "at::Dimname" so I'm hoping that the namespacing helps out with understandability

nairbv replied:
that helps, I guess we can keep thinking about it

nairbv commented on aten/src/ATen/TensorNames.h:
```diff
@@ -0,0 +1,64 @@
+// Let's say we wish to check if tensor.names[-1] matches other.names[-1].
+// None (in tensor) cannot match A (in other) because if the None were refined
+// to A, `tensor` would have duplicate names [A, A]. Therefore we need to check
+// tensor.names [A, None] for the existence of A.
```
I like having this definition/explanation. I know it might seem obvious from the name but it might also be helpful to define 'unify' as well.

zou3519 replied:
Done

nairbv commented on aten/src/ATen/TensorNames.h:
```diff
@@ -0,0 +1,64 @@
+// tensor: Tensor[A, None]
+// other: Tensor[A]
+//
+// Let's say we wish to check if tensor.names[-1] matches other.names[-1].
```
it seems a bit funny to me to use -1 for other since it only has one dim, but I guess that makes sense since we're usually going from the right?

zou3519 replied:
Yup, because we're going from the right I wanted to write this example as going from the right.

zou3519 committed:
Update on "Improve name inference API by introducing a TensorName helper struct"


Motivation
============

Before this PR, a core problem with writing name inference rules was
that each rule needed to handle misalignment by themselves. A misaligned
name occurs when we are matching None with a non-None name, but the
non-None name already exists in the first tensor.

For example, `A` is misaligned in `Tensor[A, None] + Tensor[None, A]`.

Each op handled this in a custom way
- align_from_right (used by broadcasting) handles misalignment
- compute_matmul_outnames checks for misalignment across batch and
feature dimensions.

We can actually codify "misalignment" into something more rigorous by
folding it into the definition of `match` and eliminate special handling
of "misalignment". That is what this PR attempts to do.

Approach
============

Definition: Two names in two tensors *match* if they are equal, or if at
least one of them is a wildcard that can be *refined* to the other name.

With this new definition, to check if two names match, we need to know
about the names list that each name came from to determine if a wildcard
can successfully be *refined* to the other name.

For example, consider the following:
```
tensor: Tensor[A, None]
other: Tensor[None, A]`
```
when unifying `tensor.names[-1]` with `other.names[-1]`, we see that
`tensor.names[-1]` is None and `other.names[-1]` is A. Then we check to
see if `tensor.names[-1]` can be refined to `A`; it can't be refined if
there is already an `A` in `tensor.names`.

Enter `TensorNames`.
A TensorName represents a Dimname associated with some DimnameList
(that came from a Tensor).

`TensorNames` is a list of such TensorName objects with some helper
functions attached.

One can perform the following operations:
- unify two `TensorName` objects
- unify two `TensorNames` objects with right alignment.

Plan
============

This PR changes `compute_matmul_outnames` to use `TensorNames` to
demonstrate how they make writing name inference rules easier. In the
future I'll convert other name inference rules to use `TensorNames` as
well.

Test Plan
- run all tests

[ghstack-poisoned]

nairbv commented on aten/src/ATen/NamedTensorUtils.cpp:
```diff
@@ -311,81 +312,31 @@ static bool are_distinct(DimnameList batch_dims, DimnameList feature_dims) {
-  // (3) concatenate the batch outnames and matrix multiply outnames.
-
-  // Step 1: Compute outnames of matrix multiplication
-  // Let N >= 2.
```
I still see `>=2` checks after this PR, should we preserve at least some aspects of the explanation in these comments?

zou3519 replied:
Sure. Most of this explanation is about the semantics of matmul (which I am assuming that a future dev who is reading the code understands) but I'll write up a concise explanation

zou3519 replied:
Done

zou3519 committed:
Update on "Improve name inference API by introducing a TensorName helper struct"


Motivation
============

Before this PR, a core problem with writing name inference rules was
that each rule needed to handle misalignment by themselves. A misaligned
name occurs when we are matching None with a non-None name, but the
non-None name already exists in the first tensor.

For example, `A` is misaligned in `Tensor[A, None] + Tensor[None, A]`.

Each op handled this in a custom way
- align_from_right (used by broadcasting) handles misalignment
- compute_matmul_outnames checks for misalignment across batch and
feature dimensions.

We can actually codify "misalignment" into something more rigorous by
folding it into the definition of `match` and eliminate special handling
of "misalignment". That is what this PR attempts to do.

Approach
============

Definition: Two names in two tensors *match* if they are equal, or if at
least one of them is a wildcard that can be *refined* to the other name.

With this new definition, to check if two names match, we need to know
about the names list that each name came from to determine if a wildcard
can successfully be *refined* to the other name.

For example, consider the following:
```
tensor: Tensor[A, None]
other: Tensor[None, A]`
```
when unifying `tensor.names[-1]` with `other.names[-1]`, we see that
`tensor.names[-1]` is None and `other.names[-1]` is A. Then we check to
see if `tensor.names[-1]` can be refined to `A`; it can't be refined if
there is already an `A` in `tensor.names`.

Enter `TensorNames`.
A TensorName represents a Dimname associated with some DimnameList
(that came from a Tensor).

`TensorNames` is a list of such TensorName objects with some helper
functions attached.

One can perform the following operations:
- unify two `TensorName` objects
- unify two `TensorNames` objects with right alignment.

Plan
============

This PR changes `compute_matmul_outnames` to use `TensorNames` to
demonstrate how they make writing name inference rules easier. In the
future I'll convert other name inference rules to use `TensorNames` as
well.

Test Plan
- run all tests

[ghstack-poisoned]
